go/libraries/doltcore/sqle/database_provider.go

```diff
@@ -61,9 +115,201 @@ func (p DoltDatabaseProvider) AllDatabases() (all []sql.Database) {
+		db, _, err := dbRevisionForBranch(ctx, srcDb, revSpec)
+		if err != nil {
+			return nil, false, err
+		}
+		return db, true, nil
+	}
+
+	if doltdb.IsValidCommitHash(revSpec) {
+		db, _, err := dbRevisionForCommit(ctx, srcDb, revSpec)
+		if err != nil {
+			return nil, false, err
+		}
+		return db, true, nil
+	}
+
+	return nil, false, nil
+}
+
+func (p DoltDatabaseProvider) RevisionDbState(ctx context.Context, revDB string) (dsess.InitialDbState, error) {
+	err := sql.ErrDatabaseNotFound.New(revDB)
+	if !strings.Contains(revDB, dbRevisionDelimiter) {
+		return dsess.InitialDbState{}, err
+	}
+
+	parts := strings.SplitN(revDB, dbRevisionDelimiter, 2)
+	dbName, revSpec := parts[0], parts[1]
+
+	candidate, ok := p.databases[dbName]
+	if !ok {
+		return dsess.InitialDbState{}, err
+	}
+	srcDb, ok := candidate.(Database)
+	if !ok {
+		return dsess.InitialDbState{}, err
+	}
+
+	if isBranch(ctx, srcDb.ddb, revSpec) {
+		// if the requested revision is a br we can
+		// write to it, otherwise make read-only
+		_, init, err := dbRevisionForBranch(ctx, srcDb, revSpec)
+		if err != nil {
+			return dsess.InitialDbState{}, err
+		}
+		return init, nil
+	}
+
+	// TODO: allow database revisions at commits
+	// much of the current database state logic depends on having
+	// a WorkingSet to reference, but Commits in the history don't
+	// have corresponding WorkingSets.
+	//if doltdb.IsValidCommitHash(revSpec) {
+	//	_, init, err := dbRevisionForCommit(ctx, srcDb, revSpec)
+	//	if err != nil {
+	//		return dsess.InitialDbState{}, err
+	//	}
+	//	return init, nil
+	//}
+
+	return dsess.InitialDbState{}, sql.ErrDatabaseNotFound.New(revDB)
+}
+
+func isBranch(ctx context.Context, ddb *doltdb.DoltDB, revSpec string) bool {
+	branches, err := ddb.GetBranches(ctx)
+	if err != nil {
+		return false
+	}
+
+	for _, br := range branches {
+		if revSpec == br.GetPath() {
+			return true
+		}
+	}
+
+	return false
+}
+
+func dbRevisionForBranch(ctx context.Context, srcDb Database, revSpec string) (Database, dsess.InitialDbState, error) {
+	branch := ref.NewBranchRef(revSpec)
+	cm, err := srcDb.ddb.ResolveCommitRef(ctx, branch)
+	if err != nil {
+		return Database{}, dsess.InitialDbState{}, err
+	}
+
+	wsRef, err := ref.WorkingSetRefForHead(branch)
+	if err != nil {
+		return Database{}, dsess.InitialDbState{}, err
+	}
+
+	ws, err := srcDb.ddb.ResolveWorkingSet(ctx, wsRef)
+	if err != nil {
+		return Database{}, dsess.InitialDbState{}, err
+	}
+
+	dbName := srcDb.Name() + dbRevisionDelimiter + revSpec
+
+	static := staticRepoState{
+		branch:          branch,
+		RepoStateWriter: srcDb.rsw,
+		RepoStateReader: srcDb.rsr,
+		DocsReadWriter:  srcDb.drw,
+	}
+	db := Database{
+		name: dbName,
+		ddb:  srcDb.ddb,
+		rsw:  static,
+		rsr:  static,
+		drw:  static,
+	}
+	init := dsess.InitialDbState{
+		Db:         db,
+		HeadCommit: cm,
+		WorkingSet: ws,
+		DbData: env.DbData{
+			Ddb: srcDb.ddb,
+			Rsw: static,
+			Rsr: static,
+			Drw: static,
+		},
+	}
```

Comment:
This is such a mess lol. We should combine the three interfaces into another interface and reference that one instead. 6 separate copies of `static` is just ridiculous.

Reply:
This is more of a consequence of the mess that is the `sqle` data model. fixing that will fix this